File: labelsuite/core/list_generator.py
"""검사 목록 생성 코어 — 레거시 LiGen(label_inspector_ligen.py)의 비즈니스 규칙을
GUI 없이 순수 함수로 이식. 다음 결함을 수정한다:

- 2/29 제조일: replace(year+3) ValueError → relativedelta로 안전 계산
- 행 단위 오류 무단 소멸(except: continue) → RowIssue로 수집·보고
- REF 날짜 오인 휴리스틱(':' 포함/40000 초과면 무조건 제외) → 앵커된 패턴만 날짜 취급
- 일본 GTIN BSC 미매칭 시 빈 값 → 품목리스트 폴백 + 경고
- GTIN 13자리 패딩 → normalize_gtin14(GTIN-14) 단일 규칙
- 국가 정보를 STANDARD 컬럼으로 출력해 검사 단계의 규격 자동 선택 지원
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from datetime import date, datetime, timedelta

import pandas as pd
from dateutil.relativedelta import relativedelta

from labelsuite.core.schema import (
    LabelRecord,
    SchemaError,
    date_format_for,
    normalize_gtin14,
)


@dataclass(frozen=True)
class SheetSpec:
    sheet: str
    columns: dict[str, int]

    def col(self, name: str) -> int:
        return self.columns[name]


@dataclass(frozen=True)
class ColumnMaps:
    schedule: SheetSpec
    product: SheetSpec
    bsc: SheetSpec

    @classmethod
    def from_config(cls, raw: dict) -> "ColumnMaps":
        def spec(key: str) -> SheetSpec:
            entry = raw[key]
            return SheetSpec(sheet=entry["sheet"],
                             columns={k: int(v) for k, v in entry["columns"].items()})
        return cls(schedule=spec("schedule"), product=spec("product"), bsc=spec("bsc"))


@dataclass
class RowIssue:
    row_index: int          # 엑셀 기준 행 번호(1-기반, 헤더 포함)
    lot: str
    severity: str           # 'warning' | 'error'
    message: str


@dataclass
class GenerationResult:
    records: list[LabelRecord] = field(default_factory=list)
    issues: list[RowIssue] = field(default_factory=list)
    selected_dates: list[date] = field(default_factory=list)

    @property
    def warning_count(self) -> int:
        return sum(1 for i in self.issues if i.severity == "warning")

    @property
    def error_count(self) -> int:
        return sum(1 for i in self.issues if i.severity == "error")


def load_input_frame(path: str, spec: SheetSpec) -> pd.DataFrame:
    """입력 엑셀 로드. 레거시와 동일하게 첫 행을 헤더로 소비한다."""
    return pd.read_excel(path, sheet_name=spec.sheet)


def _coerce_date(value: object) -> date | None:
    if pd.isna(value):
        return None
    try:
        return pd.to_datetime(value).date()
    except (ValueError, TypeError):
        return None


def extract_available_dates(schedule_df: pd.DataFrame, colmaps: ColumnMaps) -> list[date]:
    """스케줄 MFG DATE 열에서 중복 제거·정렬된 날짜 목록."""
    column = schedule_df.iloc[:, colmaps.schedule.col("mfg")]
    dates = {d for d in (_coerce_date(v) for v in column.dropna()) if d is not None}
    return sorted(dates)


def compute_exp_date(mfg: date, shelf_life_months: int = 36) -> date:
    """유효기한 = 제조일 + 유효기간 - 1일. relativedelta가 2/29를 월말로 보정한다."""
    return mfg + relativedelta(months=shelf_life_months) - timedelta(days=1)


# 셀 전체가 날짜(시각 선택 포함)로만 이루어진 경우만 매칭 — 부분 일치 금지
_DATE_ONLY_PATTERNS = [
    r"\d{4}-\d{1,2}-\d{1,2}",
    r"\d{1,2}/\d{1,2}/\d{4}",
    r"\d{4}/\d{1,2}/\d{1,2}",
    r"\d{1,2}-\d{1,2}-\d{4}",
    r"\d{4}\.\d{1,2}\.\d{1,2}",
    r"\d{1,2}\.\d{1,2}\.\d{4}",
]
_DATE_ONLY_RE = re.compile(
    r"^(?:" + "|".join(_DATE_ONLY_PATTERNS) + r")(?:[ T]\d{2}:\d{2}(?::\d{2})?)?$"
)


def clean_ref(raw: object) -> tuple[str, str | None]:
    """REF 셀 정리. 반환: (ref, 경고 또는 None).

    날짜로 판정하는 경우에만 REF를 비운다:
    - 값 자체가 datetime/Timestamp 인스턴스
    - 문자열 전체가 날짜(+시각) 패턴에 매칭
    - 숫자 타입이면서 엑셀 날짜 시리얼 범위(40000~60000)의 정수
    레거시의 "':' 포함 시 무조건 제외"와 "문자열 float>40000 제외"는 정상 REF를
    지우는 오판정이므로 제거했다.
    """
    if pd.isna(raw):
        return "", None
    if isinstance(raw, (datetime, pd.Timestamp)):
        return "", f"REF 셀이 날짜 값({raw})이라 비웠습니다."
    if isinstance(raw, (int, float)) and not isinstance(raw, bool):
        if float(raw).is_integer() and 40000 <= int(raw) <= 60000:
            return "", f"REF 셀이 엑셀 날짜 시리얼({int(raw)})로 보여 비웠습니다."
        text = format(int(raw), "d") if float(raw).is_integer() else str(raw)
        return text, None
    text = str(raw).strip()
    if not text or text in ("nan", "NaT"):
        return "", None
    if _DATE_ONLY_RE.match(text):
        return "", f"REF 셀이 날짜 문자열({text!r})이라 비웠습니다."
    return text, None


def _standard_for_country(country: str, country_standard_map: dict[str, str]) -> str | None:
    if country in country_standard_map:
        return country_standard_map[country] or None
    return country_standard_map.get("*") or None


def generate_list(
    schedule_df: pd.DataFrame,
    product_df: pd.DataFrame | None,
    bsc_df: pd.DataFrame | None,
    selected_dates: set[date] | list[date],
    colmaps: ColumnMaps,
    country_standard_map: dict[str, str] | None = None,
    shelf_life_months: int = 36,
) -> GenerationResult:
    """선택 날짜의 스케줄 행으로 검사 목록을 생성한다."""
    country_standard_map = country_standard_map or {}
    wanted = set(selected_dates)
    result = GenerationResult(selected_dates=sorted(wanted))
    if not wanted:
        return result

    sched = colmaps.schedule.columns

    # 조회 테이블 사전 구축 (레거시는 행마다 DataFrame 필터를 2회씩 수행)
    product_gtin_by_pn: dict[object, object] = {}
    if product_df is not None:
        pcols = colmaps.product.columns
        for _, prow in product_df.iterrows():
            pn_val = prow.iloc[pcols["pn"]]
            if pd.notna(pn_val) and pn_val not in product_gtin_by_pn:
                product_gtin_by_pn[pn_val] = prow.iloc[pcols["gtin"]]

    bsc_by_pn: dict[object, tuple[object, object]] = {}
    if bsc_df is not None:
        bcols = colmaps.bsc.columns
        for _, brow in bsc_df.iterrows():
            pn_val = brow.iloc[bcols["pn"]]
            if pd.notna(pn_val) and pn_val not in bsc_by_pn:
                bsc_by_pn[pn_val] = (brow.iloc[bcols["ref"]], brow.iloc[bcols["gtin"]])

    grouped: dict[date, list[LabelRecord]] = {d: [] for d in sorted(wanted)}

    for idx, row in schedule_df.iterrows():
        excel_row = idx + 2  # 헤더 1행 + 1-기반
        try:
            mfg = _coerce_date(row.iloc[sched["mfg"]])
            if mfg is None or mfg not in wanted:
                continue

            lot = str(row.iloc[sched["lot"]]).strip() if pd.notna(row.iloc[sched["lot"]]) else ""
            products = str(row.iloc[sched["products"]]).strip() if pd.notna(row.iloc[sched["products"]]) else ""
            pn_raw = row.iloc[sched["pn"]]
            pn = str(pn_raw).strip() if pd.notna(pn_raw) else ""
            country = str(row.iloc[sched["country"]]).strip() if pd.notna(row.iloc[sched["country"]]) else ""

            ref_base, ref_warning = clean_ref(row.iloc[sched["ref"]])
            if ref_warning:
                result.issues.append(RowIssue(excel_row, lot, "warning", ref_warning))

            standard = _standard_for_country(country, country_standard_map)
            date_fmt = date_format_for(standard)
            exp = compute_exp_date(mfg, shelf_life_months)
            mfg_str = mfg.strftime(date_fmt)
            exp_str = exp.strftime(date_fmt)

            is_japan = country == "일본"
            bsc_hit = bsc_by_pn.get(pn_raw) if pd.notna(pn_raw) else None

            # REF: 일본이면 BSC 리스트 우선, 그 외/미매칭은 스케줄 J열 값
            ref = ref_base
            if is_japan:
                if bsc_hit is not None and pd.notna(bsc_hit[0]):
                    ref = str(bsc_hit[0]).strip()
                else:
                    result.issues.append(RowIssue(
                        excel_row, lot, "warning",
                        f"일본 행의 REF를 BSC 리스트에서 찾지 못해 스케줄 값({ref_base!r})을 사용합니다."))

            # GTIN: 일본은 BSC, 그 외는 품목리스트. 일본 미매칭 시 품목리스트 폴백.
            gtin_raw: object = None
            if is_japan:
                if bsc_hit is not None and pd.notna(bsc_hit[1]):
                    gtin_raw = bsc_hit[1]
                elif pd.notna(pn_raw) and pn_raw in product_gtin_by_pn:
                    gtin_raw = product_gtin_by_pn[pn_raw]
                    result.issues.append(RowIssue(
                        excel_row, lot, "warning",
                        "일본 행의 GTIN을 BSC 리스트에서 찾지 못해 품목리스트 값으로 대체했습니다."))
            else:
                if pd.notna(pn_raw) and pn_raw in product_gtin_by_pn:
                    gtin_raw = product_gtin_by_pn[pn_raw]

            try:
                gtin = normalize_gtin14(gtin_raw)
            except SchemaError as exc:
                gtin = str(gtin_raw).strip()
                result.issues.append(RowIssue(excel_row, lot, "warning",
                                              f"GTIN 정규화 실패: {exc}"))
            if not gtin:
                result.issues.append(RowIssue(
                    excel_row, lot, "warning",
                    "GTIN을 찾지 못했습니다. 입력 파일(품목/BSC 리스트)을 확인하세요."))

            grouped[mfg].append(LabelRecord(
                lot=lot, products=products, pn=pn, ref=ref,
                mfg_date=mfg_str, exp_date=exp_str, gtin=gtin, standard=standard))
        except Exception as exc:  # 행 단위 실패는 기록하고 계속 — 무단 소멸 금지
            result.issues.append(RowIssue(excel_row, "", "error",
                                          f"행 처리 실패: {exc}"))

    for d in sorted(grouped):
        result.records.extend(grouped[d])
    return result

File: csharp/src/LabelSuite.Core/ListGenerator.cs
// 검사 목록 생성 코어 — 파이썬 core/list_generator.py 포팅 (버그 수정 포함).
using System.Globalization;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using ClosedXML.Excel;

namespace LabelSuite.Core;

public sealed record SheetSpec(string Sheet, IReadOnlyDictionary<string, int> Columns)
{
    public int Col(string name) => Columns[name];
}

public sealed record ColumnMaps(SheetSpec Schedule, SheetSpec Product, SheetSpec Bsc)
{
    public static ColumnMaps FromConfig(JsonObject raw)
    {
        SheetSpec Spec(string key)
        {
            var entry = raw[key]!.AsObject();
            var columns = new Dictionary<string, int>();
            foreach (var (name, value) in entry["columns"]!.AsObject())
                columns[name] = value!.GetValue<int>();
            return new SheetSpec(entry["sheet"]!.GetValue<string>(), columns);
        }
        return new ColumnMaps(Spec("schedule"), Spec("product"), Spec("bsc"));
    }
}

public sealed record RowIssue(int RowIndex, string Lot, string Severity, string Message);

public sealed class GenerationResult
{
    public List<LabelRecord> Records { get; } = [];
    public List<RowIssue> Issues { get; } = [];
    public List<DateOnly> SelectedDates { get; init; } = [];
    public int WarningCount => Issues.Count(i => i.Severity == "warning");
    public int ErrorCount => Issues.Count(i => i.Severity == "error");
}

/// <summary>입력 엑셀 시트를 0-기반 인덱스 접근 가능한 행 배열로 로드.</summary>
public sealed class InputFrame
{
    public List<object?[]> Rows { get; } = [];   // 데이터 행 (헤더 제외)
    public List<int> RowNumbers { get; } = [];   // 엑셀 실제 행 번호

    public static InputFrame Load(string path, SheetSpec spec)
    {
        using var workbook = new XLWorkbook(path);
        if (!workbook.Worksheets.Contains(spec.Sheet))
            throw new SchemaException($"시트 '{spec.Sheet}'를 찾을 수 없습니다: {Path.GetFileName(path)}");
        var sheet = workbook.Worksheet(spec.Sheet);
        var frame = new InputFrame();
        var maxColumn = spec.Columns.Values.Max() + 1;
        foreach (var row in sheet.RowsUsed().Skip(1))   // 첫 행은 헤더 (레거시와 동일)
        {
            var cells = new object?[maxColumn];
            for (var i = 0; i < maxColumn; i++)
            {
                var cell = row.Cell(i + 1);
                cells[i] = cell.Value.Type switch
                {
                    XLDataType.Blank => null,
                    XLDataType.DateTime => cell.GetDateTime(),
                    XLDataType.Number => cell.GetDouble(),
                    _ => cell.GetString(),
                };
            }
            frame.Rows.Add(cells);
            frame.RowNumbers.Add(row.RowNumber());
        }
        return frame;
    }
}

public static class ListGenerator
{
    /// <summary>유효기한 = 제조일 + 유효기간(개월) - 1일. AddMonths가 2/29를 월말로 보정.</summary>
    public static DateOnly ComputeExpDate(DateOnly mfg, int shelfLifeMonths = 36) =>
        mfg.AddMonths(shelfLifeMonths).AddDays(-1);

    private static readonly Regex DateOnlyRe = new(
        @"^(?:\d{4}-\d{1,2}-\d{1,2}|\d{1,2}/\d{1,2}/\d{4}|\d{4}/\d{1,2}/\d{1,2}" +
        @"|\d{1,2}-\d{1,2}-\d{4}|\d{4}\.\d{1,2}\.\d{1,2}|\d{1,2}\.\d{1,2}\.\d{4})" +
        @"(?:[ T]\d{2}:\d{2}(?::\d{2})?)?$", RegexOptions.Compiled);

    /// <summary>REF 셀 정리 — 명백한 날짜만 비우고 경고를 남긴다 (레거시 과잉 휴리스틱 제거).</summary>
    public static (string Ref, string? Warning) CleanRef(object? raw)
    {
        switch (raw)
        {
            case null: return ("", null);
            case DateTime dt: return ("", $"REF 셀이 날짜 값({dt:yyyy-MM-dd})이라 비웠습니다.");
            case double d when d == Math.Floor(d) && d is >= 40000 and <= 60000:
                return ("", $"REF 셀이 엑셀 날짜 시리얼({(long)d})로 보여 비웠습니다.");
            case double d:
                return (d == Math.Floor(d)
                    ? ((long)d).ToString(CultureInfo.InvariantCulture)
                    : d.ToString(CultureInfo.InvariantCulture), null);
        }
        var text = Convert.ToString(raw, CultureInfo.InvariantCulture)?.Trim() ?? "";
        if (text.Length == 0 || text is "nan" or "NaT") return ("", null);
        if (DateOnlyRe.IsMatch(text)) return ("", $"REF 셀이 날짜 문자열('{text}')이라 비웠습니다.");
        return (text, null);
    }

    private static DateOnly? CoerceDate(object? value) => value switch
    {
        DateTime dt => DateOnly.FromDateTime(dt),
        double d and >= 20000 and <= 80000 => DateOnly.FromDateTime(DateTime.FromOADate(d)),
        string s when DateTime.TryParse(s, CultureInfo.InvariantCulture,
                                        DateTimeStyles.None, out var parsed)
            => DateOnly.FromDateTime(parsed),
        _ => null,
    };

    public static List<DateOnly> ExtractAvailableDates(InputFrame schedule, ColumnMaps maps)
    {
        var column = maps.Schedule.Col("mfg");
        return schedule.Rows
            .Select(r => CoerceDate(r[column]))
            .Where(d => d is not null)
            .Select(d => d!.Value)
            .Distinct().OrderBy(d => d).ToList();
    }

    private static string? StandardForCountry(string country,
                                              IReadOnlyDictionary<string, string> map)
    {
        if (map.TryGetValue(country, out var standard))
            return string.IsNullOrEmpty(standard) ? null : standard;
        return map.TryGetValue("*", out var fallback) && !string.IsNullOrEmpty(fallback)
            ? fallback : null;
    }

    private static string Text(object? value) =>
        Convert.ToString(value, CultureInfo.InvariantCulture)?.Trim() ?? "";

    public static GenerationResult Generate(
        InputFrame schedule, InputFrame? product, InputFrame? bsc,
        ISet<DateOnly> selectedDates, ColumnMaps maps,
        IReadOnlyDictionary<string, string> countryStandardMap,
        int shelfLifeMonths = 36)
    {
        var result = new GenerationResult
        { SelectedDates = selectedDates.OrderBy(d => d).ToList() };
        if (selectedDates.Count == 0) return result;

        var sched = maps.Schedule.Columns;

        // 조회 테이블 사전 구축 — PN은 문자열 비교
        var productGtin = new Dictionary<string, object?>();
        if (product is not null)
        {
            var pcols = maps.Product.Columns;
            foreach (var row in product.Rows)
            {
                var pn = Text(row[pcols["pn"]]);
                if (pn.Length > 0 && !productGtin.ContainsKey(pn))
                    productGtin[pn] = row[pcols["gtin"]];
            }
        }
        var bscByPn = new Dictionary<string, (object? Ref, object? Gtin)>();
        if (bsc is not null)
        {
            var bcols = maps.Bsc.Columns;
            foreach (var row in bsc.Rows)
            {
                var pn = Text(row[bcols["pn"]]);
                if (pn.Length > 0 && !bscByPn.ContainsKey(pn))
                    bscByPn[pn] = (row[bcols["ref"]], row[bcols["gtin"]]);
            }
        }

        var grouped = selectedDates.OrderBy(d => d)
            .ToDictionary(d => d, _ => new List<LabelRecord>());

        for (var i = 0; i < schedule.Rows.Count; i++)
        {
            var row = schedule.Rows[i];
            var excelRow = schedule.RowNumbers[i];
            try
            {
                var mfg = CoerceDate(row[sched["mfg"]]);
                if (mfg is null || !selectedDates.Contains(mfg.Value)) continue;

                var lot = Text(row[sched["lot"]]);
                var products = Text(row[sched["products"]]);
                var pn = Text(row[sched["pn"]]);
                var country = Text(row[sched["country"]]);

                var (refBase, refWarning) = CleanRef(row[sched["ref"]]);
                if (refWarning is not null)
                    result.Issues.Add(new RowIssue(excelRow, lot, "warning", refWarning));

                var standard = StandardForCountry(country, countryStandardMap);
                var dateFormat = standard == "BSC" ? "yyyy.MM.dd" : "yyyy-MM-dd";
                var exp = ComputeExpDate(mfg.Value, shelfLifeMonths);

                var isJapan = country == "일본";
                bscByPn.TryGetValue(pn, out var bscHit);
                var hasBscHit = bscByPn.ContainsKey(pn);

                // REF: 일본이면 BSC 우선, 미매칭은 스케줄 값 + 경고
                var refValue = refBase;
                if (isJapan)
                {
                    if (hasBscHit && Text(bscHit.Ref).Length > 0)
                        refValue = Text(bscHit.Ref);
                    else
                        result.Issues.Add(new RowIssue(excelRow, lot, "warning",
                            $"일본 행의 REF를 BSC 리스트에서 찾지 못해 스케줄 값('{refBase}')을 사용합니다."));
                }

                // GTIN: 일본은 BSC, 그 외는 품목리스트. 일본 미매칭 시 품목리스트 폴백.
                object? gtinRaw = null;
                if (isJapan)
                {
                    if (hasBscHit && bscHit.Gtin is not null && Text(bscHit.Gtin).Length > 0)
                        gtinRaw = bscHit.Gtin;
                    else if (productGtin.TryGetValue(pn, out var fallback))
                    {
                        gtinRaw = fallback;
                        result.Issues.Add(new RowIssue(excelRow, lot, "warning",
                            "일본 행의 GTIN을 BSC 리스트에서 찾지 못해 품목리스트 값으로 대체했습니다."));
                    }
                }
                else if (productGtin.TryGetValue(pn, out var value))
                {
                    gtinRaw = value;
                }

                string gtin;
                try { gtin = Schema.NormalizeGtin14(gtinRaw); }
                catch (SchemaException ex)
                {
                    gtin = Text(gtinRaw);
                    result.Issues.Add(new RowIssue(excelRow, lot, "warning",
                        $"GTIN 정규화 실패: {ex.Message}"));
                }
                if (gtin.Length == 0)
                    result.Issues.Add(new RowIssue(excelRow, lot, "warning",
                        "GTIN을 찾지 못했습니다. 입력 파일(품목/BSC 리스트)을 확인하세요."));

                grouped[mfg.Value].Add(new LabelRecord(
                    lot, products, pn, refValue,
                    mfg.Value.ToString(dateFormat, CultureInfo.InvariantCulture),
                    exp.ToString(dateFormat, CultureInfo.InvariantCulture),
                    gtin, standard));
            }
            catch (Exception ex)   // 행 단위 실패는 기록하고 계속 — 무단 소멸 금지
            {
                result.Issues.Add(new RowIssue(excelRow, "", "error", $"행 처리 실패: {ex.Message}"));
            }
        }

        foreach (var date in grouped.Keys.OrderBy(d => d))
            result.Records.AddRange(grouped[date]);
        return result;
    }
}
